### PR TITLE
Remove PageConfig

### LINKED ISSUE #4: Remove domain from PageConfig
We can grab the `Host` header from the `Request` directly as part of `getRenderedRoute` and not need this aspect of the config. This would reduce the config to a `newtype` over `(a -> position)`, which we could keep or just pass that function directly to `withPage`.

### PR DESCRIPTION
**Pull request by pbrisbin**
- Use request `Host` for determining domain

  Fixes #4.

- Remove `PageConfig`

  Now that it's a single function we can inject it directly. Removing
  `entityPage` also means we can remove a dependency on `persistent` from our
  library target.

### CODE REVIEW

File: src/Yesod/Page/RenderedRoute.hs [hunk 1]
```diff
@@ -74,11 +67,15 @@ getRenderedRoute domain = do
...
 
   pure $ RenderedRoute
-    { renderedRouteDomain = domain
+    { renderedRouteDomain = decodeUtf8 <$> mHost
     , renderedRoutePath = path
     , renderedRouteQuery = query
     }
 
+requestHost :: MonadHandler m => m (Maybe ByteString)
+requestHost = runMaybeT $ MaybeT (lookupHeader "X-Forwarded-Host") <|> MaybeT
```
Comment (pbrisbin) [inline src/Yesod/Page/RenderedRoute.hs]: This could be written without `MaybeT`... :man_shrugging:

File: test/Spec.hs [hunk 2]
```diff
@@ -213,6 +215,33 @@ main = do
...
+      runNoLoggingT . runDB' $ do
+        deleteAssignments
+        replicateM_ 6 . insert $ SomeAssignment 1 2 now
+      request $ do
+        setUrl SomeR
+        addRequestHeader ("Host", "example.com")
+        addGetParam "teacherId" "1"
+        addGetParam "limit" "2"
+      link <- getLink "next"
+      liftIO $ link `shouldSatisfy` ("example.com" `T.isPrefixOf`)
```
Comment (pbrisbin) [inline test/Spec.hs]: So this raises some questions for me that we were ignoring when this was an opaque user-provided value...

Was it expected that `PageConfig{baseDomain}` be configured as a value including scheme? I think our pagination links should include a scheme, right? Or at least prefix the host value with `//`, which (at least in `src` attributes) tells a browser to use the same scheme as the current document. Also, will `(X-Forwarded-)Host` include a port (when it's not `80`), or do we need to find that elsewhere?

Comment (eborden) [reply test/Spec.hs]: By schema do you mean protocol? Or are you speaking about general URL components?

I'm assuming you are talking about general URL components, since you mentioned port numbers. I think it might be fine to punt on port numbers for now, but that is an open question which probably deserves an issue.

I'm cool with `//` since it coops the behavior of the client. The idea is that this is HATEAOS. A client need only extract and call a `Link` header in order to traverse to the desired hyper media artifact. They need not understand the link to utilize it.

Comment (pbrisbin) [reply test/Spec.hs]: > By schema do you mean protocol? Or are you speaking about general URL components?

I meant the `http://` part of a URL. This link does call the thing I was referring to `scheme`: https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#Definition. Though the `//` part is actually part of `authority` (which includes `host`). Words :man_shrugging: 

> I'm assuming you are talking about general URL components, since you mentioned port numbers

I mentioned port numbers because that's just another case of the same concern. After reading that wikipedia definition, I'm now realizing `userinfo` is too.

> A client need only extract and call a `Link` header in order to traverse to the desired hyper media artifact. They need not understand the link to utilize it.

Sorry, my question was assuming that guiding goal. Same page, for sure :+1:. I'm raising the flag that it's not being met. I'm also not saying it has to be met or trying to pin down what it means to meet it either, just want to have the discussion so we know what use-cases we're supporting or not at this stage.

What assumptions can we make about who our clients are and what it means to "call" these links. Do we have a standard client we're considering (e.g. `XMLHttpRequest()`)? Should our links work with `Network.HTTP.Client.parseUrl`?

> The idea is that this is HATEOAS

All the `HATEOAS` documentation I found shows relative links in their examples. No `scheme`, or `authority` at all. Is that an option?

To summarize:

Given a request to `https://admin:pass@example.com:8080/foos?limit=1`, our `Link`s should be:

1. `example.com/foo/bar?limit=1&position=y` (PR as-is)
1. `https://admin:pass@example.com:8080/foo/bar?limit=1&position=y`
1. `//example.com/foo?bar?limit=1&position=y`
1. `/foo/bar?limit=1&position=y`

Definitely give a separate "for now" and "ideally" answer too, if they're different.

For reference, the GitHub API seems to do (2). I'd bet if you hit their API on an Enterprise instance with `userinfo` and `port`, you'd get them back in the pagination links.

> I'm cool with `//` since it coops the behavior of the client

I like that reasoning, but if you take it to its logical conclusion don't you end up at (4)? I like (4) because this PR would be strictly removal.

Comment (eborden) [reply test/Spec.hs]: > All the HATEOAS documentation I found shows relative links in their examples. No scheme, or authority at all. Is that an option?

That would simplify things.
